codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -347,54 +324,78 @@ private static void initAttrs(Channel channel, Map<AttributeKey<?>, Object> attr
     }
 
     final class Http2StreamChannel extends AbstractHttp2StreamChannel implements ChannelFutureListener {
-        boolean onStreamClosedFired;
 
-        /**
-         * {@code true} if stream is in {@link Http2MultiplexCodec#channelsToFireChildReadComplete}.
-         */
+        /** {@code true} after the first HEADERS frame has been written **/
+        boolean firstFrameWritten;
+
+        /** {@code true} if a close without an error was initiated **/
+        boolean streamClosedWithoutError;
+
+        /** {@code true} if stream is in {@link Http2MultiplexCodec#channelsToFireChildReadComplete}. **/
         boolean inStreamsToFireChildReadComplete;
 
-        Http2StreamChannel(Channel parentChannel) {
-            super(parentChannel);
+        Http2StreamChannel(Channel parentChannel, Http2Stream2 stream) {
+            super(parentChannel, stream);
+            stream.managedState(this);
         }
 
         @Override
         protected void doClose() throws Exception {
-            if (!onStreamClosedFired && isStreamIdValid(streamId())) {
-                Http2StreamFrame resetFrame = new DefaultHttp2ResetFrame(Http2Error.CANCEL).streamId(streamId());
-                writeFromStreamChannel(resetFrame, true);
+            if (!streamClosedWithoutError && isStreamIdValid(stream().id())) {
+                Http2StreamFrame resetFrame = new DefaultHttp2ResetFrame(Http2Error.CANCEL).stream(stream());
+                writeFromStreamChannel(resetFrame, ctx.newPromise(), true);
             }
             super.doClose();
         }
 
         @Override
-        protected void doWrite(Object msg) {
-
+        protected void doWrite(Object msg, ChannelPromise childPromise) {
             if (msg instanceof Http2StreamFrame) {
-                Http2StreamFrame frame = (Http2StreamFrame) msg;
-                ChannelPromise promise = ctx.newPromise();
-                if (isStreamIdValid(frame.streamId())) {
+               // Http2StreamFrame frame = (Http2StreamFrame) msg;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)